tests/run-pass/catch_panic.rs

```diff
@@ -0,0 +1,50 @@
+use std::panic::catch_unwind;
+use std::cell::Cell;
+
+thread_local! {
+    static MY_COUNTER: Cell<usize> = Cell::new(0);
+    static DROPPED: Cell<bool> = Cell::new(false);
+}
+
+struct DropTester;
+
+impl Drop for DropTester {
+    fn drop(&mut self) {
+        DROPPED.with(|c| {
+            c.set(true);
+        });
+    }
+}
+
+fn do_panic_counter() {
+    // If this gets leaked, it will be easy to spot
+    // in Miri's leak report
+    let _string = "LEAKED FROM do_panic_counter".to_string();
+
+    // When we panic, this should get dropped during unwinding
+    let _drop_tester = DropTester;
+
+    // Check for bugs in Miri's panic implementation.
+    // If do_panic_counter() somehow gets called more than once,
+    // we'll generate a different panic message
+    let old_val = MY_COUNTER.with(|c| {
+        let val = c.get();
+        c.set(val + 1);
+        val
+    });
+    panic!(format!("Hello from panic: {:?}", old_val));
+}
+
+fn main() {
+    let res = catch_unwind(do_panic_counter);
```

Comment:
Could you add one more level of indirection here, to check that we unwind all stack frames properly and not just the one causing the panic?

Like
```rust
catch_unwind(|| {
let _string = "LEAKED FROM closure".to_string();
do_panic_counter()
})
```